### ISSUE TITLE
Allow storage layout overrides

### ISSUE THREAD

**Issue by beerefine** [ethereum/solidity#12920]
## Abstract

TLDR: Implement a flag similarly to Vyper's "--storage-layout-file" flag. See: https://githubhot.com/repo/vyperlang/vyper/issues/2572

## Motivation

Having the ability to control storage layout (slots) brings interesting benefits.
For one, controlling the exact storage location provides greater stability and consistency between contracts and different versions.

For example, the `owner()` view method may compile to different bytecodes depending on the underlying storage slots, even though their implementation logic is essentially the same. It forces contracts to implement the same logic again and again, deploying duplicated code, accumulating junk at expense of deployers & entire blockchain infrastructure.

As a consequence - this will open the possibility to access storage by name rather than by number, in a JSON-like manner.

For example, if I have the following contract:

```
import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract MyCode1 is Ownable, ERC20 {
    constructor() ERC20("", "") { }
}
contract MyCode2 is ERC20, Ownable {
    constructor() ERC20("", "") { }
}
```

Currently, the slot for `owner` depends on the order of inheritance, as shown in the build-info output: (slot 0 vs. 5 in the examples above)

```
"storageLayout": {
  "storage": [
    {
      "contract": "contracts/test2.sol:MyCode1",
      "label": "_owner",
      "slot": "0",
    },
    {
      "contract": "contracts/test2.sol:MyCode1",
      "label": "_balances",
      "slot": "1",
    },

...
...

"storageLayout": {
  "storage": [
    {
      "contract": "contracts/test2.sol:MyCode2",
      "label": "_balances",
      "slot": "0",
    },
    ...
    {
      "contract": "contracts/test2.sol:MyCode2",
      "label": "_owner",
      "slot": "5",
    }
  ],
...
}
```

## Specification

Instead of using slot numbers starting from 0, 1, 2, ... I suggest using the storage name in ASCII, with zero padding to 32-bytes.
With name-as-key, "owner" will be at slot 0x6f776e6572000000000000000000000000000000000000000000000000000000.

Helper methods can be implemented, e.g. `readKey(bytes32 key)`, `writeKey(bytes32 key)`, etc. for cross-contract storage access.

## Backwards Compatibility

It is not expected to break existing ABIs.

**Comment by beerefine** [comment]
I tried patching `libsolidity/interface/StorageLayout.cpp` by simply changing `_slot.str()` to `_var.name()` which compiled but I didn't get around to debugging the issue, I assume hex-encoding is required.

**Comment by hrkrshnn** [comment]
> For example, the balanceOf() or owner() view methods may compile to different bytecodes because the underlying storage slots differ, even though their implementation logic is essentially the same.

The implementation logic in this case should be different, shouldn't it? @beerefine 

> this will open the possibility to access storage by name rather than by number

Not sure if this is what you are referring to, but you can access the slot of a variable in inline assembly via `.slot` member. For example `owner.slot`.

**Comment by hrkrshnn** [comment]
About a feature to override storage slots, there is a huge downside to this: having this feature would open up possibilities for storage collisions. Even extra checks wouldn't be able to catch things like collisions between a random variable `uint owner` and a mapping `mapping (address => uint) balanceOf` (currently you can ignore this issue, but it becomes a serious concern if users can set the slot of `owner` to particular slots).

Solidity guarantees that the storage slots would remain unchanged across compiler versions for the same contract. If upgradability is something you have in mind, I believe, OpenZeppelin has built some tools that check if two contracts have the same storage slots (using the compiler output `--storage-slots`). Regrading this feature in Vyper,  Vyper doesn't have this guarantee for storage slots remaining consistent across versions, and likely need this additional feature for certain upgradable contract.

Also, getting storage slots from an external file is probably a bad idea, from a practical standpoint. Since each of the frameworks differ in how they send external flags into `solc`, I think it's risky to have contracts fetch this information from a file. A better alternative is to add a syntax for it in the language. There should be other issues for this.

**Comment by beerefine** [comment]
> The implementation logic in this case should be different, shouldn't it? @beerefine

I wasn't clear; I gave two different examples of methods, not that both of these should have the same logic. I edited my original post for clarity.

 
> Not sure if this is what you are referring to, but you can access the slot of a variable in inline assembly via `.slot` member. For example `owner.slot`.

I know it is possible with some hacking around, but my use case fits better as a compiler flag/syntax.


> About a feature to override storage slots, there is a huge downside to this: having this feature would open up possibilities for storage collisions. Even extra checks wouldn't be able to catch things like collisions between a random variable `uint owner` and a mapping `mapping (address => uint) balanceOf` (currently you can ignore this issue, but it becomes a serious concern if users can set the slot of `owner` to particular slots).

I don't see how. in my vision, owner will go to the specific slot 0x"owner"0000... (or 0x00..00"owner") but even if you control the slots completely - why wouldn't you be able to, if as you said this can be done via .slot anyway? technically even existing mappings have collisions, they will just be as rare as private key collisions.

> Solidity guarantees that the storage slots would remain unchanged across compiler versions for the same contract. If upgradability is something you have in mind, I believe, OpenZeppelin has built some tools that check if two contracts have the same storage slots (using the compiler output `--storage-slots`). Regrading this feature in Vyper, Vyper doesn't have this guarantee for storage slots remaining consistent across versions, and likely need this additional feature for certain upgradable contract.

I see, but my intention is not to replace existing contracts and possibly break upgrades - it is that future contracts will be able to use predictable slots. (and with syntax you can have the same contract produce same slots across compiler versions)

> Also, getting storage slots from an external file is probably a bad idea, from a practical standpoint. Since each of the frameworks differ in how they send external flags into `solc`, I think it's risky to have contracts fetch this information from a file. A better alternative is to add a syntax for it in the language. There should be other issues for this.

I agree; syntax sounds better, or perhaps a flag/syntax for the specific case of slot == ASCII of var name.
